src/nni_manager/training_service/pai/paiTrainingService.ts

```diff
@@ -478,41 +357,219 @@ class PAITrainingService implements TrainingService {
+            // dataDir
+            this.paiTrialConfig.dataDir,
+            // outputDir
+            this.paiTrialConfig.outputDir,
+            // codeDir
+            `$PAI_DEFAULT_FS_URI${hdfsCodeDir}`,
+            // PAI Task roles
+            paiTaskRoles,
+            // Add Virutal Cluster
+            this.paiTrialConfig.virtualCluster === undefined ? 'default' : this.paiTrialConfig.virtualCluster.toString()
+        );
+
+        // Step 2. Upload code files in codeDir onto HDFS
+        try {
+            await HDFSClientUtility.copyDirectoryToHdfs(trialLocalTempFolder, hdfsCodeDir, this.hdfsClient);
+        } catch (error) {
+            this.log.error(`PAI Training service: copy ${this.paiTrialConfig.codeDir} to HDFS ${hdfsCodeDir} failed, error is ${error}`);
+            throw new Error(error.message);
+        }
+
+        // Step 3. Submit PAI job via Rest call
+        // Refer https://github.com/Microsoft/pai/blob/master/docs/rest-server/API.md for more detail about PAI Rest API
+        const submitJobRequest: request.Options = {
+            uri: `http://${this.paiClusterConfig.host}/rest-server/api/v1/user/${this.paiClusterConfig.userName}/jobs`,
+            method: 'POST',
+            json: true,
+            body: paiJobConfig,
+            headers: {
+                'Content-Type': 'application/json',
+                Authorization: `Bearer ${this.paiToken}`
+            }
+        };
+        request(submitJobRequest, (error: Error, response: request.Response, body: any) => {
+            if (error || response.statusCode >= 400) {
+                this.log.error(`PAI Training service: Submit trial ${trialJobId} to PAI Cluster failed,
+                    error: ${error}, response: ${response}`);
+                trialJobDetail.status = 'FAILED';
+                deferred.reject(error ? error.message : `Submit trial failed, http code: ${response.statusCode}`);
+            } else {
+                trialJobDetail.submitTime = Date.now();
+                deferred.resolve(true);
+            }
+        });
+
+        return deferred.promise;
+    }
+
     private generateSequenceId(): number {
         if (this.nextTrialSequenceId === -1) {
             this.nextTrialSequenceId = getInitTrialSequenceId();
         }
 
         return this.nextTrialSequenceId++;
     }
-    
+
+    private async statusCheckingLoop(): Promise<void> {
+        while (!this.stopping) {
+            await this.updatePaiToken();
+            await this.paiJobCollector.retrieveTrialStatus(this.paiToken, this.paiClusterConfig);
+            const restServer: PAIJobRestServer = component.get(PAIJobRestServer);
+            if (restServer.getErrorMessage) {
+                throw new Error(restServer.getErrorMessage);
+            }
+            await delay(3000);
+        }
+    }
+
+    private async submitJobLoop(): Promise<void> {
+        while (!this.stopping) {
+            while (this.jobQueue.length > 0) {
```

Comment:
If does not check this.stopping within this while loop, it might induce a situation that the experiment cannot be stopped, because there are still a lot of trials waiting for submission.

Reply:
thanks, updated.